README.md
[![Serverless Nextjs Tencent Cloud](https://img.serverlesscloud.cn/2020224/1582553715762-next.js_%E9%95%BF.png)](http://serverless.com)

&nbsp;

# 腾讯云 Next.js Serverless Component

## 简介

**腾讯云[Next.js](https://github.com/zeit/next.js) 组件** - 通过使用[**Tencent Serverless Framework**](https://github.com/serverless/components/tree/cloud) , 基于云上 Serverless 服务（如 API 网关、云函数等），实现“0”配置，便捷开发，极速部署采用 Next.js 框架的网页应用，Next.js 组件支持丰富的配置扩展，提供了目前便捷实用，开发成本低的网页应用项目的开发/托管能力。

特性介绍：

- [x] **按需付费** - 按照请求的使用量进行收费，没有请求时无需付费
- [x] **"0"配置** - 只需要关心项目代码，之后部署即可，Serverless Framework 会搞定所有配置。
- [x] **极速部署** - 部署速度快，仅需几秒，部署你的整个应用。
- [x] **实时日志** - 通过实时日志的输出查看业务状态，便于直接在云端开发应用。
- [x] **云端调试** - 可在云端直接进行项目调试，从而避免本地环境的差异。
- [x] **便捷协作** - 通过云端控制台的状态信息和部署日志，方便进行多人协作开发。

## 快速开始

0. [**准备**](#0-准备)
1. [**安装**](#1-安装)
1. [**配置**](#2-配置)
1. [**部署**](#3-部署)
1. [**开发调试**](#4-开发调试)
1. [**查看状态**](#5-查看部署状态)
1. [**移除**](#6-移除)

更多资源：

- [**账号配置**](#账号配置)
- [**架构说明**](#架构说明)
- [**更多组件**](#更多组件)
- [**FAQ**](#FAQ)

### 1. 安装

通过 npm 全局安装 [serverless cli](https://github.com/serverless/serverless)

```bash
$ npm install -g serverless
```

### 2. 创建

通过如下命令和模板链接，快速创建一个 Next.js 应用：

```bash
$ serverless init nextjs-starter --name example
$ cd example
```

### 3. 部署

在 `serverless.yml` 文件所在的项目根目录，运行以下指令，将会弹出二维码，直接扫码授权进行部署：

```
serverless deploy
```

> **说明**：如果鉴权失败，请参考 [权限配置](https://cloud.tencent.com/document/product/1154/43006) 进行授权。

### 4. 配置

nextjs 组件支持 0 配置部署，也就是可以直接通过配置文件中的默认值进行部署。但你依然可以修改更多可选配置来进一步开发该 nextjs 项目。

以下是 nextjs 组件的 `serverless.yml`配置示例：

```yml
# serverless.yml
component: nextjs # (必填) 组件名称，此处为nextjs
name: nextjsDemo # (必填) 实例名称
org: orgDemo # (可选) 用于记录组织信息，默认值为您的腾讯云账户 appid
app: appDemo # (可选) 该 next.js 应用名称
stage: dev # (可选) 用于区分环境信息，默认值是 dev

inputs:
  src:
    src: ./
    exclude:
      - .env
  functionName: nextjsDemo
  region: ap-guangzhou
  runtime: Nodejs10.15
  apigatewayConf:
    protocols:
      - http
      - https
    environment: release
```

- 点此查看[更多配置及说明](/docs/configure.md)

### 5. 开发调试

部署了 Next.js 应用后，可以通过开发调试能力对该项目进行二次开发，从而开发一个生产应用。在本地修改和更新代码后，不需要每次都运行 `serverless deploy` 命令来反复部署。你可以直接通过 `serverless dev` 命令对本地代码的改动进行检测和自动上传。

可以通过在 `serverless.yml`文件所在的目录下运行 `serverless dev` 命令开启开发调试能力。

`serverless dev` 同时支持实时输出云端日志，每次部署完毕后，对项目进行访问，即可在命令行中实时输出调用日志，便于查看业务情况和排障。

### 6. 查看部署状态

在`serverless.yml`文件所在的目录下，通过如下命令查看部署状态：

```
$ serverless info
```

### 6. 移除

在`serverless.yml`文件所在的目录下，通过以下命令移除部署通过以下命令移除部署的 API 网关，移除后该组件会对应删除云上部署时所创建的所有相关资源。

```bash
$ serverless remove
```

和部署类似，支持通过 `serverless remove --debug` 命令查看移除过程中的实时日志信息。

### 账号配置

当前默认支持 CLI 扫描二维码登录，如您希望配置持久的环境变量/秘钥信息，也可以本地创建 `.env` 文件

```bash
$ touch .env # 腾讯云的配置信息
```

在 `.env` 文件中配置腾讯云的 SecretId 和 SecretKey 信息并保存

如果没有腾讯云账号，可以在此 [注册新账号](https://cloud.tencent.com/register)。

如果已有腾讯云账号，可以在 [API 密钥管理](https://console.cloud.tencent.com/cam/capi) 中获取 `SecretId` 和`SecretKey`.

```text
# .env
TENCENT_SECRET_ID=123
TENCENT_SECRET_KEY=123
```

> 注意：海外 ip 登录时，需要在`.env`文件中添加`SERVERLESS_PLATFORM_VENDOR=tencent` ，使 serverless 默认使用 tencent 组件

## 架构说明

Next.js 组件将在腾讯云账户中使用到如下 Serverless 服务：

- [x] **API 网关** - API 网关将会接收外部请求并且转发到 SCF 云函数中。
- [x] **SCF 云函数** - 云函数将承载 Next.js 应用。
- [x] **CAM 访问控制** - 该组件会创建默认 CAM 角色用于授权访问关联资源。
- [x] **COS 对象存储** - 为确保上传速度和质量，云函数压缩并上传代码时，会默认将代码包存储在特定命名的 COS 桶中
- [x] **SSL 证书服务** - 如果你在 yaml 文件中配置了 `apigatewayConf.customDomains` 字段，需要做自定义域名绑定并开启 HTTPS 时，也会用到证书管理服务和域名服务。Serverless Framework 会根据已经备案的域名自动申请并配置 SSL 证书。

## 更多组件

可以在 [Serverless Components](https://github.com/serverless/components) 仓库中查询更多组件的信息。

## 项目迁移

如果项目使用了自定义 Node.js 服务，比如 express 或者 koa，你需要做如下改造工作。

### 自定义 express 服务

如果你的 Next.js 项目本身运行就是基于 `express` 自定义服务的，那么你需要在项目中自定义入口文件 `sls.js`，需要参考你的服务启动文件进行修改，以下是一个模板文件：

```js
const express = require('express')
const next = require('next')

// not report route for custom monitor
const noReportRoutes = ['/_next', '/static', '/favicon.ico']

async function createServer() {
  const app = next({ dev: false })
  const handle = app.getRequestHandler()

  await app.prepare()

  const server = express()
  server.all('*', (req, res) => {
    noReportRoutes.forEach((route) => {
      if (req.path.indexOf(route) !== -1) {
        req.__SLS_NO_REPORT__ = true
      }
    })
    return handle(req, res)
  })

  // define binary type for response
  // if includes, will return base64 encoded, very useful for images
  server.binaryTypes = ['*/*']

  return server
}

module.exports = createServer
```

### 自定义 koa 服务

如果你的项目使用的是 Koa 作为 Node.js 服务，需要在项目中自定义入口文件 `sls.js`，需要参考你的服务启动文件进行修改，以下是一个模板文件：

```js
const Koa = require('koa')
const next = require('next')

async function createServer() {
  const app = next({ dev: false })
  const handle = app.getRequestHandler()

  const server = new Koa()
  server.use((ctx) => {
    ctx.status = 200
    ctx.respond = false
    ctx.req.ctx = ctx

    return handle(ctx.req, ctx.res)
  })

  // define binary type for response
  // if includes, will return base64 encoded, very useful for images
  server.binaryTypes = ['*/*']

  return server
}

module.exports = createServer
```

## 自定义监控

当在部署 Next.js 应用时，如果 `serverless.yml` 中未指定 `role`，默认会尝试绑定 `QCS_SCFExcuteRole`，并且开启自定义监控，帮助用户收集应用监控指标。对于为自定义入口文件的项目，会默认上报除含有 `/_next`、`/static` 和 `/favicon.ico` 的路由。如果你想自定义上报自己的路由性能，那么可以自定义 `sls.js` 入口文件，对于无需上报的路由，在 express 服务的 `req` 对象上添加 `__SLS_NO_REPORT__` 属性值为 `true` 即可。比如：

```js
server.get('/no-report', (req, res) => {
  req.__SLS_NO_REPORT__ = true
  return handle(req, res)
})
```

那么用户在访问 `GET /no-report` 路由时，就不会上报自定义监控指标。

## 文件上传

[文件上传教程](https://github.com/serverless-components/tencent-nextjs/tree/master/docs/upload.md)

## License

MIT License

Copyright (c) 2020 Tencent Cloud, Inc.


CHANGELOG.md
## [0.2.1](https://github.com/serverless-components/tencent-nextjs/compare/v0.2.0...v0.2.1) (2021-01-26)


### Bug Fixes

* change cos access by policy ([677c524](https://github.com/serverless-components/tencent-nextjs/commit/677c524868b736ea9bc839e1e235c4d46d58f780))

# [0.2.0](https://github.com/serverless-components/tencent-nextjs/compare/v0.1.10...v0.2.0) (2021-01-26)


### Features

* support apigw base64 encode ([444567e](https://github.com/serverless-components/tencent-nextjs/commit/444567e129da21d3440849e037df49e6925b6f76))

## [0.1.10](https://github.com/serverless-components/tencent-nextjs/compare/v0.1.9...v0.1.10) (2020-12-21)


### Bug Fixes

* update remove flow ([#18](https://github.com/serverless-components/tencent-nextjs/issues/18)) ([a343607](https://github.com/serverless-components/tencent-nextjs/commit/a34360709ee03c35027a349956a7ed59b56464ed))

## [0.1.9](https://github.com/serverless-components/tencent-nextjs/compare/v0.1.8...v0.1.9) (2020-12-15)


### Bug Fixes

* update to deployment to serial flow ([f96518a](https://github.com/serverless-components/tencent-nextjs/commit/f96518a4f2f3254f7b7c2c74e8cb3e383e176595))

## [0.1.8](https://github.com/serverless-components/tencent-nextjs/compare/v0.1.7...v0.1.8) (2020-10-20)


### Bug Fixes

* support replace deployment and cos assets ([85e9b7c](https://github.com/serverless-components/tencent-nextjs/commit/85e9b7c7dea22fb1eb991de4aca5cfcedc6035a7))

## [0.1.7](https://github.com/serverless-components/tencent-nextjs/compare/v0.1.6...v0.1.7) (2020-10-12)


### Bug Fixes

* support all parameters for apigw ([5e22cf3](https://github.com/serverless-components/tencent-nextjs/commit/5e22cf3595c1adfcf84de008d7af94bfc1bc4184))

## [0.1.6](https://github.com/serverless-components/tencent-nextjs/compare/v0.1.5...v0.1.6) (2020-10-10)


### Bug Fixes

* support all configs for scf ([52b8f51](https://github.com/serverless-components/tencent-nextjs/commit/52b8f51b1cebcf8d16a658454256f4f6014f7e0b))

## [0.1.5](https://github.com/serverless-components/tencent-nextjs/compare/v0.1.4...v0.1.5) (2020-09-29)


### Bug Fixes

* support customize sls entry file ([eb5d6b3](https://github.com/serverless-components/tencent-nextjs/commit/eb5d6b3c2406ce7a225296d3e541f1efda029dd5))

## [0.1.4](https://github.com/serverless-components/tencent-nextjs/compare/v0.1.3...v0.1.4) (2020-09-25)


### Bug Fixes

* support koa server ([fd37c09](https://github.com/serverless-components/tencent-nextjs/commit/fd37c09412b5e123840e841289404055814d5bec))

## [0.1.3](https://github.com/serverless-components/tencent-nextjs/compare/v0.1.2...v0.1.3) (2020-09-23)


### Bug Fixes

* update deps ([85ce058](https://github.com/serverless-components/tencent-nextjs/commit/85ce05849638ef0a318915aa39cdbc5a6ae88133))

## [0.1.2](https://github.com/serverless-components/tencent-nextjs/compare/v0.1.1...v0.1.2) (2020-09-03)


### Bug Fixes

* update deps ([b98c0f6](https://github.com/serverless-components/tencent-nextjs/commit/b98c0f653b58fb9a45f1a6baaa5e0a70e47d3b12))

## [0.1.1](https://github.com/serverless-components/tencent-nextjs/compare/v0.1.0...v0.1.1) (2020-09-01)


### Bug Fixes

* update tencent-component-toolkit ([7642596](https://github.com/serverless-components/tencent-nextjs/commit/7642596228f2972cd857096f4d30a31483175175))

# [0.1.0](https://github.com/serverless-components/tencent-nextjs/compare/v0.0.14...v0.1.0) (2020-08-28)


### Bug Fixes

* update deps for uniform error message ([b2a0ea4](https://github.com/serverless-components/tencent-nextjs/commit/b2a0ea423e89a1ec4694bb8fae11b881412b6f36))


### Features

* optimize deploy flow and update deps ([6c34a4d](https://github.com/serverless-components/tencent-nextjs/commit/6c34a4d1d8f74be5f55aa7a1659f5786b366771b))

## [0.0.14](https://github.com/serverless-components/tencent-nextjs/compare/v0.0.13...v0.0.14) (2020-08-26)


### Bug Fixes

* add no report for favicon ([28484c4](https://github.com/serverless-components/tencent-nextjs/commit/28484c48e6011004df81b1177dd19c03f4b077a1))
* apigw custom domain update bug ([9f856b5](https://github.com/serverless-components/tencent-nextjs/commit/9f856b565e67a4656d041e5f0d277f759f3c01ff))
* apigw metics x timestamp ([d68a581](https://github.com/serverless-components/tencent-nextjs/commit/d68a581e02d267871191d16d010c34f1798b0735))
* cache http server ([8e01009](https://github.com/serverless-components/tencent-nextjs/commit/8e01009823fc1682f75f17e7061b9934b03d5f38))
* change staticCdn to staticConf ([e481a94](https://github.com/serverless-components/tencent-nextjs/commit/e481a9492f1c9f7dba0c168fce0be73fa61cd0f4))
* component name ([2f45508](https://github.com/serverless-components/tencent-nextjs/commit/2f455081075379f209600a200b44bd91a50d2658))
* lambdaArn value ([0f63bda](https://github.com/serverless-components/tencent-nextjs/commit/0f63bdae851f6b3ef14c141ee23fd4d2507f54bb))
* read bucket and object from srcOriginal ([c57ae19](https://github.com/serverless-components/tencent-nextjs/commit/c57ae19dd1457d6554fa0b6fc92a6984fc474262))
* static cos has suffix appid ([1bf7400](https://github.com/serverless-components/tencent-nextjs/commit/1bf74008c729ca927e9fb2a201340606f9e8d6e3))
* support apigw serviceTimeout ([3afd507](https://github.com/serverless-components/tencent-nextjs/commit/3afd507a4b72bf5113322ead4d950df6a3a717f2))
* support eip config ([fa0358b](https://github.com/serverless-components/tencent-nextjs/commit/fa0358b85f2ca9af1abe86e1c815c7d9e3ef63ac))
* throw error when no temp secrets ([fe0b6b0](https://github.com/serverless-components/tencent-nextjs/commit/fe0b6b061bec839db8022d3d0f36fee89dd7a146))
* traffic zero display bug ([6e37e82](https://github.com/serverless-components/tencent-nextjs/commit/6e37e824e6e6009dc1fb952ce3e4ed1b38eb8564))
* typo ([2d26ef7](https://github.com/serverless-components/tencent-nextjs/commit/2d26ef732a97f3e56ef9eb96d6b849fea76eaff4))

src/serverless.js
const { Component } = require('@serverless/core')
const { Scf, Apigw, Cns, Cam, Metrics, Cos, Cdn } = require('tencent-component-toolkit')
const { TypeError } = require('tencent-component-toolkit/src/utils/error')
const {
  deepClone,
  uploadCodeToCos,
  getDefaultProtocol,
  prepareInputs,
  prepareStaticCosInputs,
  prepareStaticCdnInputs
} = require('./utils')
const CONFIGS = require('./config')

class ServerlessComopnent extends Component {
  getCredentials() {
    const { tmpSecrets } = this.credentials.tencent

    if (!tmpSecrets || !tmpSecrets.TmpSecretId) {
      throw new TypeError(
        'CREDENTIAL',
        'Cannot get secretId/Key, your account could be sub-account and does not have the access to use SLS_QcsRole, please make sure the role exists first, then visit https://cloud.tencent.com/document/product/1154/43006, follow the instructions to bind the role to your account.'
      )
    }

    return {
      SecretId: tmpSecrets.TmpSecretId,
      SecretKey: tmpSecrets.TmpSecretKey,
      Token: tmpSecrets.Token
    }
  }

  getAppId() {
    return this.credentials.tencent.tmpSecrets.appId
  }

  async deployFunction(credentials, inputs, regionList) {
    if (!inputs.role) {
      try {
        const camClient = new Cam(credentials)
        const roleExist = await camClient.CheckSCFExcuteRole()
        if (roleExist) {
          inputs.role = 'QCS_SCFExcuteRole'
        }
      } catch (e) {
        // no op
      }
    }

    const outputs = {}
    const appId = this.getAppId()

    const funcDeployer = async (curRegion) => {
      const code = await uploadCodeToCos(this, appId, credentials, inputs, curRegion)
      const scf = new Scf(credentials, curRegion)
      const tempInputs = {
        ...inputs,
        code
      }
      const scfOutput = await scf.deploy(deepClone(tempInputs))
      outputs[curRegion] = {
        functionName: scfOutput.FunctionName,
        runtime: scfOutput.Runtime,
        namespace: scfOutput.Namespace
      }

      this.state[curRegion] = {
        ...(this.state[curRegion] ? this.state[curRegion] : {}),
        ...outputs[curRegion]
      }

      // default version is $LATEST
      outputs[curRegion].lastVersion = scfOutput.LastVersion
        ? scfOutput.LastVersion
        : this.state.lastVersion || '$LATEST'

      // default traffic is 1.0, it can also be 0, so we should compare to undefined
      outputs[curRegion].traffic =
        scfOutput.Traffic !== undefined
          ? scfOutput.Traffic
          : this.state.traffic !== undefined
          ? this.state.traffic
          : 1

      if (outputs[curRegion].traffic !== 1 && scfOutput.ConfigTrafficVersion) {
        outputs[curRegion].configTrafficVersion = scfOutput.ConfigTrafficVersion
        this.state.configTrafficVersion = scfOutput.ConfigTrafficVersion
      }

      this.state.lastVersion = outputs[curRegion].lastVersion
      this.state.traffic = outputs[curRegion].traffic
    }

    for (let i = 0; i < regionList.length; i++) {
      const curRegion = regionList[i]
      await funcDeployer(curRegion)
    }
    this.save()
    return outputs
  }

  // try to add dns record
  async tryToAddDnsRecord(credentials, customDomains) {
    try {
      const cns = new Cns(credentials)
      for (let i = 0; i < customDomains.length; i++) {
        const item = customDomains[i]
        if (item.domainPrefix) {
          await cns.deploy({
            domain: item.subDomain.replace(`${item.domainPrefix}.`, ''),
            records: [
              {
                subDomain: item.domainPrefix,
                recordType: 'CNAME',
                recordLine: '默认',
                value: item.cname,
                ttl: 600,
                mx: 10,
                status: 'enable'
              }
            ]
          })
        }
      }
    } catch (e) {
      console.log('METHOD_tryToAddDnsRecord', e.message)
    }
  }

  async deployApigateway(credentials, inputs, regionList) {
    if (inputs.isDisabled) {
      return {}
    }

    const getServiceId = (instance, region) => {
      const regionState = instance.state[region]
      return inputs.serviceId || (regionState && regionState.serviceId)
    }

    const deployTasks = []
    const outputs = {}
    regionList.forEach((curRegion) => {
      const apigwDeployer = async () => {
        const apigw = new Apigw(credentials, curRegion)

        const oldState = this.state[curRegion] || {}
        const apigwInputs = {
          ...inputs,
          oldState: {
            apiList: oldState.apiList || [],
            customDomains: oldState.customDomains || []
          }
        }
        // different region deployment has different service id
        apigwInputs.serviceId = getServiceId(this, curRegion)
        const apigwOutput = await apigw.deploy(deepClone(apigwInputs))
        outputs[curRegion] = {
          serviceId: apigwOutput.serviceId,
          subDomain: apigwOutput.subDomain,
          environment: apigwOutput.environment,
          url: `${getDefaultProtocol(inputs.protocols)}://${apigwOutput.subDomain}/${
            apigwOutput.environment
          }${apigwInputs.endpoints[0].path}`
        }

        if (apigwOutput.customDomains) {
          // TODO: need confirm add cns authentication
          if (inputs.autoAddDnsRecord === true) {
            // await this.tryToAddDnsRecord(credentials, apigwOutput.customDomains)
          }
          outputs[curRegion].customDomains = apigwOutput.customDomains
        }
        this.state[curRegion] = {
          created: true,
          ...(this.state[curRegion] ? this.state[curRegion] : {}),
          ...outputs[curRegion],
          apiList: apigwOutput.apiList
        }
      }
      deployTasks.push(apigwDeployer())
    })

    await Promise.all(deployTasks)

    this.save()
    return outputs
  }

  // deploy static to cos, and setup cdn
  async deployStatic(credentials, inputs, region) {
    const { zipPath } = this.state
    const appId = this.getAppId()
    const deployStaticOutpus = {}

    if (zipPath) {
      console.log(`Deploy static for ${CONFIGS.compFullname} application`)
      // 1. deploy to cos
      const { staticCosInputs, bucket } = await prepareStaticCosInputs(
        this,
        inputs,
        appId,
        zipPath,
        region
      )

      const cos = new Cos(credentials, region)
      const cosOutput = {
        region
      }
      // flush bucket
      if (inputs.cosConf.replace) {
        await cos.flushBucketFiles(bucket)
      }
      for (let i = 0; i < staticCosInputs.length; i++) {
        const curInputs = staticCosInputs[i]
        console.log(`Starting deploy directory ${curInputs.src} to cos bucket ${curInputs.bucket}`)
        const deployRes = await cos.deploy(curInputs)
        cosOutput.cosOrigin = `${curInputs.bucket}.cos.${region}.myqcloud.com`
        cosOutput.bucket = deployRes.bucket
        cosOutput.url = `https://${curInputs.bucket}.cos.${region}.myqcloud.com`
        console.log(`Deploy directory ${curInputs.src} to cos bucket ${curInputs.bucket} success`)
      }
      deployStaticOutpus.cos = cosOutput

      // 2. deploy cdn
      if (inputs.cdnConf) {
        const cdn = new Cdn(credentials)
        const cdnInputs = await prepareStaticCdnInputs(this, inputs, cosOutput.cosOrigin)
        console.log(`Starting deploy cdn ${cdnInputs.domain}`)
        const cdnDeployRes = await cdn.deploy(cdnInputs)
        const protocol = cdnInputs.https ? 'https' : 'http'
        const cdnOutput = {
          domain: cdnDeployRes.domain,
          url: `${protocol}://${cdnDeployRes.domain}`,
          cname: cdnDeployRes.cname
        }
        deployStaticOutpus.cdn = cdnOutput

        console.log(`Deploy cdn ${cdnInputs.domain} success`)
      }

      console.log(`Deployed static for ${CONFIGS.compFullname} application successfully`)

      return deployStaticOutpus
    }

    return null
  }

  async deploy(inputs) {
    console.log(`Deploying ${CONFIGS.compFullname} App...`)

    const credentials = this.getCredentials()

    // 对Inputs内容进行标准化
    const { regionList, functionConf, apigatewayConf } = await prepareInputs(
      this,
      credentials,
      inputs
    )

    // 部署函数 + API网关
    const outputs = {}
    if (!functionConf.code.src) {
      outputs.templateUrl = CONFIGS.templateUrl
    }

    let apigwOutputs
    const functionOutputs = await this.deployFunction(
      credentials,
      functionConf,
      regionList,
      outputs
    )
    // support apigatewayConf.isDisabled
    if (apigatewayConf.isDisabled !== true) {
      apigwOutputs = await this.deployApigateway(credentials, apigatewayConf, regionList, outputs)
    } else {
      this.state.apigwDisabled = true
    }

    // optimize outputs for one region
    if (regionList.length === 1) {
      const [oneRegion] = regionList
      outputs.region = oneRegion
      outputs['scf'] = functionOutputs[oneRegion]
      if (apigwOutputs) {
        outputs['apigw'] = apigwOutputs[oneRegion]
      }
    } else {
      outputs['scf'] = functionOutputs
      if (apigwOutputs) {
        outputs['apigw'] = apigwOutputs
      }
    }

    // start deploy static cdn
    if (inputs.staticConf) {
      const staticOutputs = {}
      const staticTasks = []
      regionList.forEach((region) => {
        const staticDeployer = async () => {
          const res = await this.deployStatic(credentials, inputs.staticConf, region)
          staticOutputs[region] = res
        }
        staticTasks.push(staticDeployer())
      })

      await Promise.all(staticTasks)

      if (regionList.length === 1) {
        const [oneRegion] = regionList
        this.state.staticConf = staticOutputs[oneRegion]
        outputs.staticConf = staticOutputs[oneRegion]
      } else {
        this.state.staticConf = staticOutputs
        outputs.staticConf = staticOutputs
      }
    }

    this.state.region = regionList[0]
    this.state.regionList = regionList

    this.state.lambdaArn = functionConf.name

    return outputs
  }

  async removeStatic() {
    // remove static
    const { region, staticConf } = this.state
    if (staticConf) {
      console.log(`Removing static config`)
      const credentials = this.getCredentials()
      // 1. remove cos
      if (staticConf.cos) {
        const cos = new Cos(credentials, region)
        await cos.remove(staticConf.cos)
      }
      // 2. remove cdn
      if (staticConf.cdn) {
        const cdn = new Cdn(credentials)
        try {
          await cdn.remove(staticConf.cdn)
        } catch (e) {
          // no op
        }
      }
      console.log(`Remove static config success`)
    }
  }

  async remove() {
    console.log(`Removing ${CONFIGS.compFullname} App...`)

    const { state } = this
    const { regionList = [] } = state

    const credentials = this.getCredentials()

    const removeHandlers = []
    for (let i = 0; i < regionList.length; i++) {
      const curRegion = regionList[i]
      const curState = state[curRegion]
      const scf = new Scf(credentials, curRegion)
      const apigw = new Apigw(credentials, curRegion)
      const handler = async () => {
        // if disable apigw, no need to remove
        if (state.apigwDisabled !== true) {
          await apigw.remove({
            created: curState.created,
            environment: curState.environment,
            serviceId: curState.serviceId,
            apiList: curState.apiList,
            customDomains: curState.customDomains
          })
        }
        await scf.remove({
          functionName: curState.functionName,
          namespace: curState.namespace
        })
      }
      removeHandlers.push(handler())
    }

    await Promise.all(removeHandlers)

    // remove static
    await this.removeStatic()

    this.state = {}
  }

  async metrics(inputs = {}) {
    console.log(`Get ${CONFIGS.compFullname} Metrics Datas...`)
    if (!inputs.rangeStart || !inputs.rangeEnd) {
      throw new TypeError(
        `PARAMETER_${CONFIGS.compName.toUpperCase()}_METRICS`,
        'rangeStart and rangeEnd are require inputs'
      )
    }
    const { region } = this.state
    if (!region) {
      throw new TypeError(
        `PARAMETER_${CONFIGS.compName.toUpperCase()}_METRICS`,
        'No region property in state'
      )
    }
    const { functionName, namespace, functionVersion } = this.state[region] || {}
    if (functionName) {
      const options = {
        funcName: functionName,
        namespace: namespace,
        version: functionVersion,
        region,
        timezone: inputs.tz
      }
      const curState = this.state[region]
      if (curState.serviceId) {
        options.apigwServiceId = curState.serviceId
        options.apigwEnvironment = curState.environment || 'release'
      }
      const credentials = this.getCredentials()
      const mertics = new Metrics(credentials, options)
      const metricResults = await mertics.getDatas(
        inputs.rangeStart,
        inputs.rangeEnd,
        Metrics.Type.All
      )
      return metricResults
    }
    throw new TypeError(
      `PARAMETER_${CONFIGS.compName.toUpperCase()}_METRICS`,
      'Function name not define'
    )
  }
}

module.exports = ServerlessComopnent
